File: ios/Classes/TiGooglemapsCircleProxy.m
/**
 * Appcelerator Titanium Mobile
 * Copyright (c) 2009-Present by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

#import "TiGooglemapsCircleProxy.h"

@implementation TiGooglemapsCircleProxy

-(GMSCircle*)circle
{
    if (_circle == nil) {
        _circle = [GMSCircle new];
        _circle.tappable = YES;
    }
    
    return _circle;
}

-(void)dealloc
{
    RELEASE_TO_NIL(_circle);
    [super dealloc];
}

#pragma mark Public APIs

-(void)setCenter:(id)args
{
    ENSURE_UI_THREAD(setCenter, args);
    [self replaceValue:args forKey:@"center" notification:NO];
    
    if(![args isKindOfClass:[NSArray class]] && ![args isKindOfClass:[NSDictionary class]]) {
        NSLog(@"[WARN] Ti.GoogleMaps: You need to specify the center either using an array or object.");
        return;
    }
    
    [[self circle] setPosition:[self positionFromPoint:args]];
}

-(void)setRadius:(id)value
{
    ENSURE_UI_THREAD(setRadius, value);
    ENSURE_TYPE(value, NSNumber);
    
    [[self circle] setRadius:[TiUtils doubleValue:value]];
    [self replaceValue:value forKey:@"radius" notification:NO];
}

-(void)setTappable:(id)value
{
    ENSURE_UI_THREAD(setTappable, value);
    ENSURE_TYPE(value, NSNumber);
    
    [[self circle] setTappable:[TiUtils boolValue:value]];
    [self replaceValue:value forKey:@"tappable" notification:NO];
}

-(void)setFillColor:(id)value
{
    ENSURE_UI_THREAD(setFillColor, value);
    
    [[self circle] setFillColor:[[TiUtils colorValue:value] _color]];
    [self replaceValue:value forKey:@"fillColor" notification:NO];
}

-(void)setStrokeColor:(id)value
{
    ENSURE_UI_THREAD(setStrokeColor, value);
    
    [[self circle] setStrokeColor:[[TiUtils colorValue:value] _color]];
    [self replaceValue:value forKey:@"strokeColor" notification:NO];
}

-(void)setStrokeWidth:(id)value
{
    ENSURE_UI_THREAD(setStrokeWidth, value);
    
    [[self circle] setStrokeWidth:[TiUtils floatValue:value def:1]];
    [self replaceValue:value forKey:@"strokeWidth" notification:NO];
}

#pragma mark Utilities

-(CLLocationCoordinate2D)positionFromPoint:(id)point
{
    if ([point isKindOfClass:[NSDictionary class]]) {
        CLLocationDegrees latitude = [TiUtils doubleValue:[point valueForKey:@"latitude"]];
        CLLocationDegrees longitude = [TiUtils doubleValue:[point valueForKey:@"longitude"]];

        return CLLocationCoordinate2DMake(latitude, longitude);
    } else if ([point isKindOfClass:[NSArray class]]) {
        CLLocationDegrees latitude = [TiUtils doubleValue:[point objectAtIndex:0]];
        CLLocationDegrees longitude = [TiUtils doubleValue:[point objectAtIndex:1]];
        
        return CLLocationCoordinate2DMake(latitude, longitude);
    }
}

@end

File: titanium-googlemaps-swift/Classes/TiGooglemapsCircleProxy.swift
/**
 * Axway Titanium
 * Copyright (c) 2018-present by Axway Appcelerator. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

import UIKit
import TitaniumKit
import GoogleMaps
import CoreLocation

@objc(TiGooglemapsCircleProxy)
public class TiGooglemapsCircleProxy: TiProxy {
  private var circleInternal: GMSCircle?
  
  private var circleInstance: GMSCircle {
    if let circle = circleInternal {
      return circle
    }
    
    let circle = GMSCircle()
    circle.isTappable = true
    circleInternal = circle
    return circle
  }
  
  private func runOnMainThread(_ block: @escaping () -> Void) {
    if Thread.isMainThread {
      block()
    } else {
      TiThreadPerformOnMainThread(block, false)
    }
  }
  
  private func coordinate(from value: Any?) -> CLLocationCoordinate2D? {
    if let dict = value as? [String: Any] {
      let latitude = TiUtils.doubleValue(dict["latitude"])
      let longitude = TiUtils.doubleValue(dict["longitude"])
      return CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    } else if let array = value as? [Any], array.count >= 2 {
      let latitude = TiUtils.doubleValue(array[1])
      let longitude = TiUtils.doubleValue(array[0])
      return CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }
    
    return nil
  }
  
  // MARK: - Public API
  
  @objc public func circle() -> GMSCircle {
    return circleInstance
  }
  
  @objc public func setCenter(_ value: Any?) {
    runOnMainThread {
      guard let coordinate = self.coordinate(from: value) else {
        NSLog("[WARN] Ti.GoogleMaps: You need to specify the center either using an array or object.")
        return
      }
      
      self.circleInstance.position = coordinate
      self.replaceValue(value ?? NSNull(), forKey: "center", notification: false)
    }
  }
  
  @objc public func setRadius(_ value: Any?) {
    runOnMainThread {
      self.circleInstance.radius = TiUtils.doubleValue(value)
      self.replaceValue(value, forKey: "radius", notification: false)
    }
  }
  
  @objc public func setTappable(_ value: Any?) {
    runOnMainThread {
      self.circleInstance.isTappable = TiUtils.boolValue(value)
      self.replaceValue(value, forKey: "tappable", notification: false)
    }
  }
  
  @objc public func setFillColor(_ value: Any?) {
    runOnMainThread {
      self.circleInstance.fillColor = TiUtils.colorValue(value)?.color
      self.replaceValue(value, forKey: "fillColor", notification: false)
    }
  }
  
  @objc public func setStrokeColor(_ value: Any?) {
    runOnMainThread {
      self.circleInstance.strokeColor = TiUtils.colorValue(value)?.color
      self.replaceValue(value, forKey: "strokeColor", notification: false)
    }
  }
  
  @objc public func setStrokeWidth(_ value: Any?) {
    runOnMainThread {
      self.circleInstance.strokeWidth = CGFloat(TiUtils.floatValue(value, def: 1))
      self.replaceValue(value, forKey: "strokeWidth", notification: false)
    }
  }
  
  @objc public func setTitle(_ value: Any?) {
    runOnMainThread {
      let title = TiUtils.stringValue(value)
      self.circleInstance.title = title
      self.replaceValue(title ?? NSNull(), forKey: "title", notification: false)
    }
  }
  
  @objc public func setZIndex(_ value: Any?) {
    runOnMainThread {
      self.circleInstance.zIndex = Int32(TiUtils.intValue(value))
      self.replaceValue(value, forKey: "zIndex", notification: false)
    }
  }
  
  @objc public func setUserData(_ value: Any?) {
    runOnMainThread {
      self.circleInstance.userData = value
      self.replaceValue(value ?? NSNull(), forKey: "userData", notification: false)
    }
  }
}
